packages/rules-engine/src/lib/runners/allIfFirst.ts

```diff
@@ -0,0 +1,18 @@
+import { type Rule, type RuleContext } from "../rule";
+
+/**
+ * Run all rules that return true for `runIf` only if the first rule returns true.
+ *
+ * @param rules The rules to run.
+ */
+export function allIfFirst<TInput, TOutput, TContext extends RuleContext<TInput, TOutput>>(
+  ...rules: Array<Rule<TInput, TOutput, TContext>>
+): Rule<TInput, TOutput, TContext> {
+  return {
+    runIf: (input) => Boolean(rules[0]?.runIf(input)),
```

Comment:
nit: I think we shouldn't have a rule, but a simple predicate based on which the rest of the rules will start.

Reply:
Example:

`allIf(isSchool, schoolRules)`

Reply:
@PetricaP This is a new runner to support the previous scenario (e.g. [here](https://github.com/ClipboardHealth/open-shifts/blob/main/curated-shifts/src/modules/shift-offers/pricing-engine/rules/fixed-rates-workplaces.ts#L30))

Reply:
It could have supported that as well with:
`allIf(isFixedWorkplace, all(basePayRule, basePayMarginRule, holidayRule, manualChargeRule))`

Reply:
This way we wouldn't need a special rule to be first in the chain.

Reply:
This slightly changes the runner type, which works on a list of rules. We can come back to this later, but in my use cases, we do need an initial rule, and only if this rule passes do we want to run the others.

Reply:
@therockstorm, what do you think about this suggestion? 

`allIf(isSchool, schoolRules)`

This is something we could definitely need in the future for the school vertical.

Reply:
Can't we just have one or the other? The fewer the primitive the better, in my opinion. It seems like  "allIf" would handle both scenarios.

Reply:
@PetricaP I think a downside of allIf() is traceability, for example, with my approach, I'll write something like:

```
export const fixedPayWorkplacePricingRule: PricingRule = {
  runIf(input): boolean {
    ///
  },
  run(context) {
    return appendOutput(context, {
      meta: {
        reasonCode: RepricingCode.FIXED_PAY,
      },
      amount: fixedPay!,// **it can be 0 if we dont need a special amount here**
    });
  },
};
```

This means we get a rule output for FIXED_PAY (even if the output is 0), while in 

`allIf(isFixedWorkplace, all(basePayRule, basePayMarginRule, holidayRule, manualChargeRule))`

We haven't stored the FIXED_PAY rule output.

Reply:
We could always include a fixed pay dummy rule if we wanted an output in the `all` so I don't think it's necessarily a downside.

allIf(isFixedWorkplace, fixedPayMarkerRule, basePayRule, basePayMarginRule, holidayRule, manualChargeRule)

I have come up with this problem in other cases as well like for RDA where we needed to duplicate the condition in `runIf` in multiple rules because we couldn't make the condition apply to a set of rules.

I think we'll keep have more and more edge cases where different sets of rules apply so we need something like the `allIf` construct.

To me ignoring the conditions in the other rules in this construct and having a convention that the first rule is somehow special seems hacky, but I don't think it's a big enough problem to worry about so it's not a blocker.

Reply:
With core libraries, people will use the constructs provided, so we want to make sure to provide constructs that are hard to misuse and that don't result in hacky code. I would say that if `allIf` works for more scenarios, we should go with it instead.

Reply:
@PetricaP, the other rules runIf are not ignored (we still use the run () when running them); it's just that those rules should run only if the **first** one pass.

I'm happy to change this, but I don't like the idea that this runner has a different structure than the others, which only need a list of rules to run.

Reply:
Wouldn't the types on the function signature be enough to let people know it's different than the other runners? We could also provide example usage and a TypeDoc comment.

This is likely my last day before paternity leave, so I'll let the two of you come to a decision.

Reply:
We're in the middle of a new experiment that will give us a better understanding of what we need (allIf, or allIfFirst)

Reply:
@yankovalera Make sure to update the examples and regenerate the example in the README with `npm run embed`, whatever you decide.

Reply:
We've added an allIf [here](https://github.com/ClipboardHealth/open-shifts/blob/main/curated-shifts/src/libs/rules-engine/runner/all-if.ts), which will be ported to core-utils in the next few weeks.

It's been helpful to have this in the new rules.